diff --git a/blocks/course_overview/block_course_overview.php b/blocks/course_overview/block_course_overview.php
index 2812cce..4dfbbad 100644
--- a/blocks/course_overview/block_course_overview.php
+++ b/blocks/course_overview/block_course_overview.php
@@ -67,6 +67,13 @@ class block_course_overview extends block_base {
         if ($updatemynumber >= 0) {
             block_course_overview_update_mynumber($updatemynumber);
         }
+        // [course_overview_ext] Begin patch.
+        // Save setting "mytimestart".
+        $updatemytimestart = optional_param("mytimestart", -1, PARAM_INT);
+        if ($updatemytimestart >= 0) {
+            set_user_preference("course_overview_timestart_for_log", $updatemytimestart);
+        }
+        // [course_overview_ext] End patch.
 
         profile_load_custom_fields($USER);
 
diff --git a/blocks/course_overview/locallib.php b/blocks/course_overview/locallib.php
index 746ccf4..4927faf 100644
--- a/blocks/course_overview/locallib.php
+++ b/blocks/course_overview/locallib.php
@@ -46,6 +46,19 @@ function block_course_overview_get_overviews($courses) {
             foreach ($modules as $fname) {
                 $fname($courses, $htmlarray);
             }
+            // [course_overview_ext] Begin patch.
+            // Show new/changed activites:
+            // It is important to call our own function AFTER the call of $fname!
+            // I need the information of the course modules in order to create the entries of $htmlarray.
+            // Update 02/09/2014: Bug - It was not possible to call the my-home page, when the block was moved above the navigation.
+            global $CFG; // Bugfix from 02/09/2014.
+            if (file_exists($CFG->dirroot . '/local/course_overview_ext/lib.php'))
+            {
+                require_once $CFG->dirroot . '/local/course_overview_ext/lib.php';
+                foreach ($courses as $course)
+                    course_overview_ext_structural_changes($course, $htmlarray);
+            }
+            // [course_overview_ext] End patch.
         }
     }
     return $htmlarray;
diff --git a/blocks/course_overview/renderer.php b/blocks/course_overview/renderer.php
index 8923322..33934f4 100644
--- a/blocks/course_overview/renderer.php
+++ b/blocks/course_overview/renderer.php
@@ -181,6 +181,17 @@ class block_course_overview_renderer extends plugin_renderer_base {
     protected function activity_display($cid, $overview) {
         $output = html_writer::start_tag('div', array('class' => 'activity_info'));
         foreach (array_keys($overview) as $module) {
+            // [course_overview_ext] Begin patch.
+            // Check, if the key have special information, which we have applied in course_overview_ext_structural_changes.
+            // After that normalize it to Moodle default. In that way, we can use core code again.
+            $onlystructural = (substr($module, 0, 5) === "COEXT"); // This key exists only, when no additional information of this module is available.
+            $action = "";
+            if ($onlystructural) {
+                $action = $overview[$module]["action"];
+                $infotext = $overview[$module]["infotext"];
+                $module = substr($module, 5);
+            }
+            // [course_overview_ext] End patch.
             $output .= html_writer::start_tag('div', array('class' => 'activity_overview'));
             $url = new moodle_url("/mod/$module/index.php", array('id' => $cid));
             $modulename = get_string('modulename', $module);
@@ -190,6 +201,18 @@ class block_course_overview_renderer extends plugin_renderer_base {
             } else {
                 $icontext .= get_string("activityoverview", 'block_course_overview', $modulename);
             }
+            // [course_overview_ext] Begin patch.
+            // Show changed activities ONLY (otherwise the message is stored in $overview[$module]).
+            if ($onlystructural) {
+                // $renew icontext. This behaviour has the advantage, that we just insert code and remain old one
+                $icontext = $this->output->pix_icon("i/" . $action, "", "local_course_overview_ext", array("class" => "iconextra"))
+                          . html_writer::link($url, $this->output->pix_icon('icon', $modulename, "mod_" . $module, array("class" => "iconlarge")))
+                          . get_string($action, "moodle", $modulename);
+                // Set $overview[$module] so that we can use the old "footer"
+                // Remember: "COEXT" was cut of before, i.e. $module is just the module name (in frankenstyle) again.
+                $overview[$module] = $infotext;
+            }
+            // [course_overview_ext] End patch.
 
             // Add collapsible region with overview text in it.
             $output .= $this->collapsible_region($overview[$module], '', 'region_'.$cid.'_'.$module, $icontext, '', true);
@@ -217,6 +240,11 @@ class block_course_overview_renderer extends plugin_renderer_base {
         $select = new single_select($url, 'mynumber', $options, block_course_overview_get_max_user_courses(), array());
         $select->set_label(get_string('numtodisplay', 'block_course_overview'));
         $output .= $this->output->render($select);
+        // [course_overview_ext] Begin patch.
+        // Start time selector for "recent activities".
+        $select = course_overview_ext_timestart_select($url);
+        $output .= $this->output->render($select);
+        // [course_overview_ext] End patch.
 
         $output .= $this->output->box_end();
         return $output;
